yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -331,14 +334,25 @@ private[spark] class ApplicationMaster(
                 finish(FinalApplicationStatus.FAILED,
                   ApplicationMaster.EXIT_REPORTER_FAILURE, "Exception was thrown " +
                     s"${failureCount} time(s) from Reporter thread.")
-
               } else {
                 logWarning(s"Reporter thread fails ${failureCount} time(s) in a row.", e)
               }
             }
           }
           try {
-            Thread.sleep(interval)
+            val numPendingAllocate = allocator.getNumPendingAllocate
+            if (numPendingAllocate > 0) {
+              currentAllocationInterval =
+                math.min(heartbeatInterval,currentAllocationInterval * 2)
+              logDebug(s"Number of pending allocations is ${numPendingAllocate}. " +
+                        "Sleeping for " + currentAllocationInterval)
+              Thread.sleep(currentAllocationInterval)
```

Comment:
One thing that is not really covered by the bug, but would be nice to add, is code to wake up this thread when a request for new executors arrives. That would help in lowering the latency for new executors after some idle time, when this loop is back to sleeping at the usual heartbeat interval.

Feel free to punt that to a separate issue, though.